rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -35,370 +37,502 @@
+// to lookup the pointer, and gain a reference to a shared_ptr if it exists.
+// This guarantees that the Data object will not be destroyed while we are using it.
+static std::mutex data_to_data_shared_ptr_map_mutex;
+static std::unordered_map<Data *, std::shared_ptr<Data>> data_to_data_shared_ptr_map;
+
+static void graph_sub_data_handler(const z_sample_t * sample, void * data);
+
+// Bundle all class members into a data struct which can be passed as a
+// weak ptr to various threads for thread-safe access without capturing
+// "this" ptr by reference.
+class Data final
 {
-  z_owned_str_t keystr = z_keyexpr_to_string(sample->keyexpr);
-  auto free_keystr = rcpputils::make_scope_exit(
-    [&keystr]() {
-      z_drop(z_move(keystr));
-    });
+public:
+  // Constructor.
+  Data(
+    std::size_t domain_id,
+    const std::string & enclave)
+  : domain_id_(std::move(domain_id)),
+    enclave_(std::move(enclave)),
+    is_shutdown_(false),
+    next_entity_id_(0),
+    nodes_({})
+  {
+    // Initialize the zenoh configuration.
+    z_owned_config_t config;
+    rmw_ret_t ret;
+    if ((ret =
+      rmw_zenoh_cpp::get_z_config(
+        rmw_zenoh_cpp::ConfigurableEntity::Session,
+        &config)) != RMW_RET_OK)
+    {
+      throw std::runtime_error("Error configuring Zenoh session.");
+    }
 
-  auto data_ptr = static_cast<Data *>(data);
-  if (data_ptr == nullptr) {
-    RMW_ZENOH_LOG_ERROR_NAMED(
-      "rmw_zenoh_cpp",
-      "[graph_sub_data_handler] Invalid data_ptr."
-    );
-    return;
-  }
+    // Check if shm is enabled.
+    z_owned_str_t shm_enabled = zc_config_get(z_loan(config), "transport/shared_memory/enabled");
+    auto always_free_shm_enabled = rcpputils::make_scope_exit(
+      [&shm_enabled]() {
+        z_drop(z_move(shm_enabled));
+      });
+
+    // Initialize the zenoh session.
+    session_ = z_open(z_move(config));
+    if (!z_session_check(&session_)) {
+      throw std::runtime_error("Error setting up zenoh session.");
+    }
+    auto close_session = rcpputils::make_scope_exit(
+      [this]() {
+        z_close(z_move(session_));
+      });
+
+    // TODO(Yadunund) Move this check into a separate thread.
+    // Verify if the zenoh router is running if configured.
+    const std::optional<uint64_t> configured_connection_attempts =
+      rmw_zenoh_cpp::zenoh_router_check_attempts();
+    if (configured_connection_attempts.has_value()) {
+      ret = RMW_RET_ERROR;
+      uint64_t connection_attempts = 0;
+      // Retry until the connection is successful.
+      while (ret != RMW_RET_OK && connection_attempts < configured_connection_attempts.value()) {
+        if ((ret = rmw_zenoh_cpp::zenoh_router_check(z_loan(session_))) != RMW_RET_OK) {
+          ++connection_attempts;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+      }
+      if (ret != RMW_RET_OK) {
+        throw std::runtime_error(
```

Comment:
include `<stdexcept>`

Reply:
Fixed in dcd2e7fab758d23e550adc1c1167800b45c08594